src/antlr/CompletionsVisitor.ts

```diff
@@ -0,0 +1,437 @@
+   * @param columns The available columns in the relation
+   */
+  addAvailableColumns(relation: RelationName, columns: ColumnName[]) {
+    this.availableColumns.push(
+      ...columns.map(name => {
+        return { relation, name };
+      })
+    );
+  }
+
+  /**
+   * Add a relation that can be used in the from clause of a query. This
+   * will either come from the global scope or a CTE.
+   *
+   * @param name The relation that contains the columns
+   * @param columns The available columns in the relation
+   */
+  addRelation(name: RelationName, columns: ColumnName[]) {
+    this.environment.set(name, columns);
+  }
+
+  /**
+   * Retrieve the columns for a relation. This method will look up the
+   * relation in this context and recursively its parent.
+   *
+   * @param relation The relation we want to read the columns from
+   * @returns The names of columns in the relation.
+   */
+  getColumns(relation: RelationName): ColumnName[] {
+    return this.environment.get(relation) ?? this.parent?.getColumns(relation) ?? [];
+  }
+
+  /**
+   * Returns the columns selected in the sub-query for this context.
+   *
+   * @returns The columns that are selected in the current sub-query.
+   */
+  getSelectedColumns() {
+    return this.selectedColumns;
+  }
+
+  /**
+   * Retrieve the columns that are available in the sub-query (i.e. columns from
+   * relations in the from clause).
+   *
+   * @param relation An optional name of the relation to filter columns by.
+   * @returns The columns available (potentially filtered by the relation name)
+   */
+  getAvailableColumns(relation?: RelationName): ColumnName[] {
+    if (relation === undefined) {
+      return this.availableColumns.map(c => c.name);
+    } else {
+      return this.availableColumns.filter(c => c.relation === relation).map(c => c.name);
+    }
+  }
+
+  private getRelations(): RelationName[] {
+    return Array.from(this.environment.keys()).concat(this.parent?.getRelations() ?? []);
+  }
+
+  /**
+   * Return all completable information about the sub-query for this context.
+   *
+   * @returns A compact representation of the completable information for the sub-query
+   */
+  getQueryInfo(): QueryInfo {
+    return {
+      availableColumns: this.availableColumns,
+      availableRelations: this.getRelations()
+    };
+  }
+}
+
+/**
+ * The part of the query to complete.
+ */
+type Scope = { type: "column" } | { type: "relation" } | { type: "dereference"; relation: string } | { type: "other" };
+
+/**
+ * The result of the completion task. Contains the completable information of the
+ * (sub-)query, the part in which to complete and applicable snippets.
+ */
+interface Result {
+  queryInfo: QueryInfo;
+  scope: Scope;
+  snippets: Snippet[];
+}
+
+/**
+ * A visitor for the SQL ast that collects all required information for the auto completion.
+ * The position for which to run completion will be identified via a special token (CARET: '<|>').
```

Comment:
I'm not really fan of changing the grammar to have this special caret... The node range should be enough, so let's not mix everything.

Let's talk on this on Monday.